distsys/resources/nestedarch.go

```diff
@@ -0,0 +1,344 @@
+// ErrNestedArchetypeSanity signifies that some internal sanity check failed regarding buffer / channel state. intended for use during panic
+var ErrNestedArchetypeSanity = errors.New("internal sanity check failed")
+
+const (
+	// nestedArchetypeTimeout is how long a nested archetype will wait on an operation before giving up
+	// this is arbitrary and hardcoded, and should be made configurable later.
+	nestedArchetypeTimeout = 100 * time.Millisecond
+)
+
+var nestedArchetypeReadReq = tla.MakeTLAString("read_req")
+var nestedArchetypeWriteReq = tla.MakeTLAString("write_req")
+var nestedArchetypePreCommitReq = tla.MakeTLAString("precommit_req")
+var nestedArchetypeAbortReq = tla.MakeTLAString("abort_req")
+var nestedArchetypeCommitReq = tla.MakeTLAString("commit_req")
+
+var nestedArchetypeReadAck = tla.MakeTLAString("read_ack")
+var nestedArchetypeWriteAck = tla.MakeTLAString("write_ack")
+var nestedArchetypePreCommitAck = tla.MakeTLAString("precommit_ack")
+var nestedArchetypeAbortAck = tla.MakeTLAString("abort_ack")
+var nestedArchetypeCommitAck = tla.MakeTLAString("commit_ack")
+
+var nestedArchetypeAborted = tla.MakeTLAString("aborted")
+
+// NestedArchetypeConstantDefs provides a quick way to include correct definitions for all the boilerplate constants
+// a resource implementation will always require.
+//
+// These definitions would satisfy roughly the following TLA+, binding each constant to its own name:
+//
+//   CONSTANTS READ_REQ, WRITE_REQ, ABORT_REQ, PRECOMMIT_REQ, COMMIT_REQ
+//   CONSTANTS READ_ACK, WRITE_ACK, ABORT_ACK, PRECOMMIT_ACK, COMMIT_ACK
+//
+var NestedArchetypeConstantDefs = distsys.EnsureMPCalContextConfigs(
+	// req tpe
+	distsys.DefineConstantValue("READ_REQ", nestedArchetypeReadReq),
+	distsys.DefineConstantValue("WRITE_REQ", nestedArchetypeWriteReq),
+	distsys.DefineConstantValue("ABORT_REQ", nestedArchetypeAbortReq),
+	distsys.DefineConstantValue("PRECOMMIT_REQ", nestedArchetypePreCommitReq),
+	distsys.DefineConstantValue("COMMIT_REQ", nestedArchetypeCommitReq),
+	// ack tpe
+	distsys.DefineConstantValue("READ_ACK", nestedArchetypeReadAck),
+	distsys.DefineConstantValue("WRITE_ACK", nestedArchetypeWriteAck),
+	distsys.DefineConstantValue("ABORT_ACK", nestedArchetypeAbortAck),
+	distsys.DefineConstantValue("PRECOMMIT_ACK", nestedArchetypePreCommitAck),
+	distsys.DefineConstantValue("COMMIT_ACK", nestedArchetypeCommitAck),
+	// aborted notification tpe
+	distsys.DefineConstantValue("ABORTED", nestedArchetypeAborted),
+)
+
+type nestedArchetypeConstantsT struct {
+	tpe tla.TLAValue
+	value tla.TLAValue
+}
+
+var nestedArchetypeConstants = nestedArchetypeConstantsT {
+	tpe: tla.MakeTLAString("tpe"),
+	value: tla.MakeTLAString("value"),
+}
+
+type nestedArchetype struct {
+	distsys.ArchetypeResourceLeafMixin
+	nestedCtxs []*distsys.MPCalContext
+	ctxErrCh chan error
+
+	sendCh chan tla.TLAValue
+	receiveCh chan tla.TLAValue
+
+	cachedErrCh chan error
```

Comment:
The channel instance itself is being cached. It's just a way to avoid running `make(chan error)` in a loop when calling the resource API... a debatable micro-optimisation, and I'm happy to remove it if it just makes things unclear.

Reply:
I see. I thought that the channel contains some cached errors. I got your point but the cached prefix in the name can be confusing. Maybe `errCh` is just fine.

Reply:
hmm... to me, that's more ambiguous.

Maybe `apiErrCh`? Then, we know it's used for the resource API. I can also add a doc comment.